### ISSUE TITLE
🐞[BUG]: Select decorator returning state from wrong store during SSR

### ISSUE THREAD

**Issue by Dav1dde** [ngxs/store#1646]
### Affected Package

`@ngxs/store`

### Is this a regression?

No

### Description

During two SSR requests (can be simulated), `@Select()` decorators return data from the wrong store, they always return data from the second request (even if processing the first).

## 🔬 Minimal Reproduction

```typescript
import { ɵCommonEngine as CommonEngine } from '@nguniversal/common/engine';
const engine = new CommonEngine(RootModule);
```

```typescript
        const options: RenderOptions = {req, bootstrap: RootModule, providers: []};
        options.url = `${req.protocol}://${(req.get('host') || '')}${req.originalUrl}`;
        options.document = document;
        options.providers = [
            {provide: REQUEST, useValue: req},
            {provide: RESPONSE, useValue: res}
        ];

        const options2 = {...options};

        Promise.all([
            engine.render(options),
            engine.render(options2)
        ]).then(([r1, r2]) => {
            res.send(r1 + r2);
        });
```

The problem is in [`SelectFactory`](https://github.com/ngxs/store/blob/master/packages/store/src/decorators/select/select-factory.ts#L13), it uses a static member to store the currently active store, with 2 requests the second overwrites the first and further processing in the first request uses the wrong store.

This makes the `@Select` decorator unusable in an SSR environment. The decorator cannot use a global storage that is not scoped to the current execution. Something like `cls-hooked` could work or the store instance needs to come from somewhere else.

## Environment

Libs:
- @angular/core version: 9
- @ngxs/store version: 3.6.2

SSR
 
For Tooling issues:
- Node version: 13
- Platform:  Linux

**Comment by arturovt** [comment]
Hey @Dav1dde 

I was looking at your issue. I have a question: did you call `engine.render()` 2 times in a single handler just for demonstrating purposes? I'm guessing that it should be 1 call to `engine.render` per 1 request like:

```ts
const fastify = require('fastify')();

fastify.get('*', (request, reply) => {
  reply.type('text/html');

  engine
    .render({
      document,
      url: request.raw.url,
      bootstrap: AppServerModule,
    })
    .then(html => {
      reply.send(html);
    });
});
```

P.S. I'm just clarify all facts.

I started also debugging and looking for a way to solve this issue. Currently, I've implemented locally a single `Map` where `id` is a currently async execution ID OR key can be current zone, so basically it looks like this:

```ts
import { NgModuleRef } from '@angular/core';
import { Store } from '@ngxs/store';
import * as asyncHooks from 'async_hooks';

class ServerSelectFactory {
  private static stores = new Map<number, Store>();

  constructor(ngModuleRef: NgModuleRef<unknown>) {
    const store: Store = ngModuleRef.injector.get(Store);
    const executionAsyncId: number = asyncHooks.executionAsyncId();

    ServerSelectFactory.stores.set(executionAsyncId, store);

    ngModuleRef.onDestroy(() => {
      ServerSelectFactory.stores.delete(executionAsyncId);
    });
  }

  static getStore(): Store {
    return this.stores.get(asyncHooks.executionAsyncId());
  }
}
```

Or:

```ts
import { Store } from '@ngxs/store';

declare const Zone: any;

class ServerSelectFactory {
  private static stores = new WeakMap<typeof Zone, Store>();

  constructor(store: Store) {
    ServerSelectFactory.stores.set(Zone.current, store);
  }

  static getStore(): Store {
    return this.stores.get(Zone.current);
  }
}
```

**Comment by Dav1dde** [comment]
@arturovt 

yeah I am calling `engine.render` twice to reproduce the issue we had in production, this matches the behavior of two concurrent requests each calling `engine.render`. You can simulate load using a benchmarking tool like apache `ab` (`ab -n 10000 -c 5 "http://localhost:4000/config-a"`) and opening an instance in your browser `http://localhost:4000/config-b` which will then show you data from the `config-a` request.

I tried myself to fix this issue with a global storage for all currently active zones/contexts like `cls-hooked`, the issue always was that the context got lost somewhere in angular. I have not tried `Zone.current`.

The current workaround for me is a custom `@Select` decorator that grabs the `store` instance from the component it is attached to (which requires you to inject the store into each component).

**Comment by arturovt** [comment]
@Dav1dde can I ask you to make a simple reproducible project which I can run locally and push it to github repo. I was struggling to reproduce thatt issue, but I understand the root cause and was trying to "get" the correct Store instance in above examples. With your example I can clarify that it works 100% correctly. Cheers!

**Comment by arturovt** [comment]
Hey, I will put this issue on hold. Unfortunately, I don't understand what means `http://localhost:4000/config-b which will then show you data from the config-a request`. Reproducible example will be appreciated.

**Comment by asaadedd** [comment]
This can be reproduced in my cloned repository: https://github.com/asaadedd/store in the app inside the integration folder. I cannot create a Stackblitz because the error is present in the SSR logs.

Check my comment for my analysis of the bug: https://github.com/ngxs/store/issues/1723#issuecomment-788019393

**Comment by markwhitfeld** [comment]
@Dav1dde and @asaadedd there is work currently happening to fix this issue.
See PR #1740 for progress. @arturovt is an absolute legend!!!

**Comment by markwhitfeld** [comment]
Hi @Dav1dde and @asaadedd, the PR from @arturovt has been merged into master, so it will be available on NPM through the `@ngxs/store@dev` version. Please could you test it out and confirm that it resolves this issue for you?

**Comment by arturovt** [comment]
This has been fixed as part of the [v3.7.4](https://github.com/ngxs/store/releases/tag/v3.7.4) release.
Please test this out and let us know if this is not resolved.
